### PR TITLE
WIP: Fix type checking of record projection by expression (#1020)

### LINKED ISSUE #1020: Illegal projection is caught only on second try
```
$ dhall repl
⊢ {foo = 1}.({foo : Bool})

{ foo = 1 }

⊢ {foo = 1}.({foo : Bool})

Error: Expression doesn't match annotation

{ foo : - Bool
        + Natural
}
```

I can reproduce this both with `v1.24.0` and on `master`.

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #1020.

### CODE REVIEW

File: dhall/src/Dhall/TypeCheck.hs [hunk 1]
```diff
@@ -802,18 +802,20 @@ typeWithA tpa = loop
...
-                        case Dhall.Set.difference keysT keysR of
-                            k : _ -> do
-                                Left (TypeError ctx e (MissingField k t))
-                            [] -> do
-                                return (Record ktsT)
+                    nf_t@(Record ktsT) -> do
+                        let projection = Record (Dhall.Map.intersection ktsR ktsT)
+
+                        t' <- loop ctx projection
+                        if Dhall.Core.judgmentallyEqual t t'
```
Comment (Gabriel439) [inline dhall/src/Dhall/TypeCheck.hs]: Did you mean for this to be `Dhall.Core.judgmentallyEqual t projection`?

For example, for the expression:

```dhall
{ foo = 1 }.{ foo : Natural }
```

.... `t` would be `{ foo : Natural }`, but `t'` would be `Type`

Comment (sjakobi) [reply dhall/src/Dhall/TypeCheck.hs]: Thanks!

Comment (Gabriel439) [reply dhall/src/Dhall/TypeCheck.hs]: You're welcome! 🙂

**Comment by Gabriel439** [comment]
@sjakobi: I put up a parallel pull request for the same fix which takes care of the more detailed error messages: https://github.com/dhall-lang/dhall-haskell/pull/1027

**Pull request closed without merging**